// @akashic/pdi-common-impl/lib/VideoPlayer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.VideoPlayer = void 0;
var trigger_1 = require("@akashic/trigger");
/**
 * ビデオ再生を行うクラス。
 *
 * ゲーム開発者は本クラスのインスタンスを直接生成すべきではない。
 */
var VideoPlayer = /** @class */ (function () {
    /**
     * `VideoPlayer` のインスタンスを生成する。
     */
    function VideoPlayer(loop) {
        this._loop = !!loop;
        this.onPlay = new trigger_1.Trigger();
        this.onStop = new trigger_1.Trigger();
        this.played = this.onPlay;
        this.stopped = this.onStop;
        this.currentVideo = undefined;
        this.volume = 1.0;
    }
    /**
     * `VideoAsset` を再生する。
     *
     * 再生後、 `this.onPlay` がfireされる。
     * @param Video 再生するビデオアセット
     */
    VideoPlayer.prototype.play = function (videoAsset) {
        this.currentVideo = videoAsset;
        this.onPlay.fire({
            player: this,
            video: videoAsset
        });
    };
    /**
     * 再生を停止する。
     *
     * 再生中でない場合、何もしない。
     * 停止後、 `this.onStop` がfireされる。
     */
    VideoPlayer.prototype.stop = function () {
        var videoAsset = this.currentVideo;
        this.onStop.fire({
            player: this,
            video: videoAsset
        });
    };
    /**
     * 音量を変更する。
     *
     * エンジンユーザが `VideoPlayer` の派生クラスを実装する場合は、
     *  このメソッドをオーバーライドして実際に音量を変更する処理を行うこと。
     *  オーバーライド先のメソッドはこのメソッドを呼びださなければならない。
     * @param volume 音量。0以上1.0以下でなければならない
     */
    VideoPlayer.prototype.changeVolume = function (volume) {
        this.volume = volume;
    };
    return VideoPlayer;
}());
exports.VideoPlayer = VideoPlayer;
